### PR TITLE
[IMP] digest: compute the KPI in batch

### PR DESCRIPTION
**Pull request by std-odoo**
Purpose
=======
Now, the KPI are computed based on their `company_id`, and not based
on the current company. If no company is set on the digest, we compute
it based on the current company.

The KPIs are computed based on their company. Most of the time, they
will all belong to the same company so we can improve the performance
by computing them in batch.

Task-2827996

### CODE REVIEW

File: addons/digest/tests/test_digest_kpi.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo.addons.digest.tests.common import TestDigestCommon
+
+
+class TestDigest(TestDigestCommon):
+
+    def test_kpi_mail_message_total_value(self):
+        # Sanity check
+        initial_values = self.all_digests.mapped('kpi_mail_message_total_value')
```
Comment (tde-banana-odoo) [inline addons/digest/tests/test_digest_kpi.py:11]: That probably highly depends on installed modules :D would reset messages in setupclass (either count existing one and check new_count = old_count + create_cnt, either unlink all existing one)

File: addons/digest/tests/test_digest_kpi.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo.addons.digest.tests.common import TestDigestCommon
+
+
+class TestDigest(TestDigestCommon):
+
+    def test_kpi_mail_message_total_value(self):
```
Comment (tde-banana-odoo) [inline addons/digest/tests/test_digest_kpi.py:9]: There is already a test for that in ``test_digest.py`` (could move ``_setup_messages`` in common by the way)

Comment (tde-banana-odoo) [reply addons/digest/tests/test_digest_kpi.py:9]: We could also use the common class in ``test_digest.py`` by the way

File: addons/digest/tests/test_digest_kpi.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+        self.all_digests.invalidate_recordset()
+
+        values = self.all_digests.mapped('kpi_mail_message_total_value')
+
+        self.assertEqual(len(set(values)), 1, 'Message Count is a cross company KPI')
+        self.assertEqual(initial_values[0] + 1, values[0])
+
+    def test_kpi_res_users_connected_value(self):
+        # Sanity check
+        initial_values = self.all_digests.mapped('kpi_res_users_connected_value')
```
Comment (tde-banana-odoo) [inline addons/digest/tests/test_digest_kpi.py:31]: Would probably have to count initial values / force initial values also

File: addons/point_of_sale/models/digest.py [hunk 4]
```diff
@@ -14,14 +14,14 @@ class Digest(models.Model):
...
-        for record in self:
-            start, end, company = record._get_kpi_compute_parameters()
-            record.kpi_pos_total_value = sum(self.env['pos.order'].search([
-                ('date_order', '>=', start),
-                ('date_order', '<', end),
-                ('state', 'not in', ['draft', 'cancel', 'invoiced']),
-                ('company_id', '=', company.id)
-            ]).mapped('amount_total'))
+
+        self._compute_company_based_kpi(
```
Comment (tde-banana-odoo) [inline addons/point_of_sale/models/digest.py:18]: There is a void addons/point_of_sale/tests/test_digest.py file to remove ^^

**Review by tde-banana-odoo — COMMENTED** [review]
Validation !

File: addons/digest/models/digest.py [hunk 5]
```diff
@@ -55,20 +56,72 @@ def _compute_available_fields(self):
...
-            total_messages = self.env['mail.message'].search_count([('create_date', '>=', start), ('create_date', '<', end), ('subtype_id', '=', discussion_subtype_id), ('message_type', 'in', ['comment', 'email'])])
-            record.kpi_mail_message_total_value = total_messages
+        self.kpi_mail_message_total_value = self.env['mail.message'].search_count([
+            ('create_date', '>=', start),
+            ('create_date', '<', end),
+            ('subtype_id', '=', discussion_subtype_id),
+            ('message_type', 'in', ('comment', 'email')),
+        ])
+
+    def _compute_company_based_kpi(self, model, field_name, date_field='create_date',
```
Comment (awa-odoo) [inline addons/digest/models/digest.py:89]: Maybe this could be moved to a "utility" section in the file as it's not actually a compute method in the strict sense? (Or rename but I don't have great suggestions).

File: addons/digest/models/digest.py [hunk 6]
```diff
@@ -55,20 +56,72 @@ def _compute_available_fields(self):
...
+                                   domain=None, sum_field=None):
+        """Generic method that computes the KPI on a given model.
+
+        :param model: Model on which we will count the number of record
```
Comment (awa-odoo) [inline addons/digest/models/digest.py:93]: typo: number of records

Note that potentially we don't count the number of records, since that's what "sum_field" is for ^^

File: addons/digest/models/digest.py [hunk 7]
```diff
@@ -55,20 +56,72 @@ def _compute_available_fields(self):
...
+            This model must have a "company_id" field
+        :param field_name: Field name on which we will write the KPI
+        :param date_field: Field used for the date range
+        :param domain: Additional domain for the search
```
Comment (awa-odoo) [inline addons/digest/models/digest.py:97]: Would not say "for the search", a bit confusing as it's a read_group.
Could also rename "additional_domain" as it's complementary to the base domain with dates / companies.

File: addons/digest/models/digest.py [hunk 8]
```diff
@@ -55,20 +56,72 @@ def _compute_available_fields(self):
...
+            ('message_type', 'in', ('comment', 'email')),
+        ])
+
+    def _compute_company_based_kpi(self, model, field_name, date_field='create_date',
+                                   domain=None, sum_field=None):
+        """Generic method that computes the KPI on a given model.
+
+        :param model: Model on which we will count the number of record
+            This model must have a "company_id" field
+        :param field_name: Field name on which we will write the KPI
```
Comment (awa-odoo) [inline addons/digest/models/digest.py:95]: digest_kpi_field ?

(To better distinguish from "sum_field" that is a field on the passed model).

File: addons/digest/models/digest.py [hunk 9]
```diff
@@ -55,20 +56,72 @@ def _compute_available_fields(self):
...
+        :param date_field: Field used for the date range
+        :param domain: Additional domain for the search
+        :param sum_field: Field to sum to obtain the KPI,
+            if None it will count the number of records
+        """
+        start, end, companies = self._get_kpi_compute_parameters()
```
Comment (awa-odoo) [inline addons/digest/models/digest.py:101]: I'm guessing it's not possible to not have start/end? For example if you wanted to compute a kpi without time constraints?

File: addons/digest/models/digest.py [hunk 10]
```diff
@@ -55,20 +56,72 @@ def _compute_available_fields(self):
...
+
+        search_domain = [
```
Comment (awa-odoo) [inline addons/digest/models/digest.py:103]: "base_domain" maybe as it's not used to search?

File: addons/digest/models/digest.py [hunk 11]
```diff
@@ -55,20 +56,72 @@ def _compute_available_fields(self):
...
+            (date_field, '>=', start),
+            (date_field, '<', end),
+        ]
+
+        if domain:
+            search_domain = expression.AND([search_domain, domain])
+
+        values = self.env[model]._read_group(
+            domain=search_domain,
+            fields=[f'{sum_field}:sum' if sum_field else 'id'],
```
Comment (awa-odoo) [inline addons/digest/models/digest.py:114]: Do you even need this part if sum_field is not defined? I thought it did that automatically.

**Review by awa-odoo — COMMENTED** [review]
Hello @std-odoo 

Left of few comments on the main digest part.
It's mostly picky stuff, looks like a nice refactoring!

Will read the other modules / tests a bit later.

Thanks & cheers!

**Review by std-odoo — COMMENTED** [review]
@awa-odoo Hi :) 

I did the changes and answered

Cheers

File: addons/digest/models/digest.py [hunk 12]
```diff
@@ -55,20 +56,72 @@ def _compute_available_fields(self):
...
-            record.kpi_res_users_connected_value = user_connected
+        self._compute_company_based_kpi(
+            'res.users',
+            'kpi_res_users_connected_value',
+            date_field='login_date',
+        )
 
     def _compute_kpi_mail_message_total_value(self):
+        start, end, __ = self._get_kpi_compute_parameters()
         discussion_subtype_id = self.env.ref('mail.mt_comment').id
```
[the remark this answers is not in the document]
Comment (awa-odoo) [reply addons/digest/models/digest.py:80]: I meant

`('subtype_id', '=', self.env.ref('mail.mt_comment').id),`

I did not count, but it does not seem to be 230 chars ^^

[the remark this answers is not in the document]
Comment (std-odoo) [reply addons/digest/models/digest.py:80]: oh sorry, I tought you talk about the domain just below, changed

File: addons/crm/tests/test_digest.py [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+        self.assertEqual(self.digest_2.kpi_crm_opportunities_won_value, 0,
+            msg='This digest is in a different company')
+        self.assertEqual(self.digest_3.kpi_crm_opportunities_won_value, 1,
+            msg='This digest has no company, should take the current one')
+
+        self.digest_3.invalidate_recordset()
+        self.assertEqual(
+            self.digest_3.with_company(self.company_2).kpi_crm_opportunities_won_value,
+            0,
+        )
```
Comment (awa-odoo) [inline addons/crm/tests/test_digest.py:68]: I feel like all those are a bit redundant as you are already testing the behavior with the second assertion.

File: addons/im_livechat/models/digest.py [hunk 14]
```diff
@@ -15,36 +15,32 @@ class Digest(models.Model):
...
-        channels = self.env['mail.channel'].search([('livechat_operator_id', '=', self.env.user.partner_id.id)])
-        for record in self:
-            start, end, company = record._get_kpi_compute_parameters()
-            domain = [
-                ('create_date', '>=', start), ('create_date', '<', end),
-                ('rated_partner_id', '=', self.env.user.partner_id.id)
-            ]
-            ratings = channels.rating_get_grades(domain)
-            record.kpi_livechat_rating_value = ratings['great'] * 100 / sum(ratings.values()) if sum(ratings.values()) else 0
+        channels = self.env['mail.channel'].search([('channel_type', '=', 'livechat')])
```
Comment (awa-odoo) [inline addons/im_livechat/models/digest.py:18]: Previous code seemingly relied on the "current partner", was that incorrect?

As we are modifying the logic here and not just "batch computing", should we make a separate commit? Maybe as a backported bugfix?

File: addons/im_livechat/models/digest.py [hunk 15]
```diff
@@ -15,36 +15,32 @@ class Digest(models.Model):
...
     def _compute_kpi_livechat_conversations_value(self):
-        for record in self:
-            start, end, company = record._get_kpi_compute_parameters()
-            record.kpi_livechat_conversations_value = self.env['mail.channel'].search_count([
-                ('channel_type', '=', 'livechat'),
-                ('livechat_operator_id', '=', self.env.user.partner_id.id),
-                ('create_date', '>=', start), ('create_date', '<', end)
-            ])
+        start, end, __ = self._get_kpi_compute_parameters()
+        self.kpi_livechat_conversations_value = self.env['mail.channel'].search_count([
```
Comment (awa-odoo) [inline addons/im_livechat/models/digest.py:32]: Well apparently it's the exact same thing for all 3 KPIs.

File: addons/im_livechat/models/digest.py [hunk 16]
```diff
@@ -15,36 +15,32 @@ class Digest(models.Model):
...
-            record.kpi_livechat_response_value = sum(
-                response['time_to_answer']
-                for response in response_time
-                if response['time_to_answer'] > 0
-            )
+        start, end, __ = self._get_kpi_compute_parameters()
+        response_time = self.env['im_livechat.report.channel'].sudo()._read_group([
+            ('start_date', '>=', start),
+            ('start_date', '<', end),
+        ], ['time_to_answer:avg'], [])
```
Comment (awa-odoo) [inline addons/im_livechat/models/digest.py:42]: Also looks like a fix, taking the average instead if the sum.

File: addons/im_livechat/tests/test_digest.py [hunk 17]
```diff
@@ -0,0 +1,57 @@
...
+        }, {
+            'rated_partner_id': other_partner.id,
+            'res_id': cls.channels[0].id,
+            'res_model_id': cls.env['ir.model']._get('mail.channel').id,
+            'consumed': True,
+            'rating': 5,
+        }])
+
+    def test_kpi_livechat_rating_value(self):
+        self.assertEqual(round(self.digest_1.kpi_livechat_rating_value, 2), 33.33)
```
Comment (awa-odoo) [inline addons/im_livechat/tests/test_digest.py:57]: Is that 33%? Can you make a quick explanation of how you get that with the initial data set? It's not super obvious ^^

File: addons/project/models/digest.py [hunk 18]
```diff
@@ -14,15 +14,13 @@ class Digest(models.Model):
...
-                ('create_date', '>=', start),
-                ('create_date', '<', end),
-                ('company_id', '=', company.id),
-                ('display_project_id', '!=', False),
-            ])
+
+        self._calculate_company_based_kpi(
+            'project.task',
+            'kpi_project_task_opened_value',
+            date_field='create_date',
```
Comment (awa-odoo) [inline addons/project/models/digest.py:21]: Maybe don't specify as it's the default?

File: addons/rating/models/rating_mixin.py [hunk 19]
```diff
@@ -130,7 +130,7 @@ def _rating_get_repartition(self, add_stats=False, domain=None):
                 otherwise, key is the value of the information (string) : either stat name (avg, total, ...) or 'repartition'
                 containing the same dict if add_stats was False.
         """
-        base_domain = expression.AND([self._rating_domain(), [('rating', '>=', 1)]])
```
Comment (awa-odoo) [inline addons/rating/models/rating_mixin.py:133]: Hmmm?

**Review by awa-odoo — COMMENTED** [review]
Hello @std-odoo 

Back to this one.
Globally the part about batching the computation is LGTM.

However there seems to be a bunch of small other functional changes or fixes that are "hidden" within the commit.
Would be better to isolate them as fixes or improvements to better distinguish the purpose of the changes.

Also you should be able to rebase and get rid of the fix in 15.0.

Thanks & cheers!

**Review by std-odoo — COMMENTED** [review]
@awa-odoo Hi :) 

I did the changes and splitted the commit

Cheers

**Comment by awa-odoo** [comment]
@robodoo r+ rebase-ff
